lib/ui/channel_buffers.dart

```diff
@@ -6,184 +6,334 @@
 
-  /// A mapping between a channel name and its associated [_RingBuffer].
-  final Map<String, _RingBuffer<_StoredMessage>?> _messages =
-    <String, _RingBuffer<_StoredMessage>?>{};
-
-  _RingBuffer<_StoredMessage> _makeRingBuffer(int size) {
-    final _RingBuffer<_StoredMessage> result = _RingBuffer<_StoredMessage>(size);
-    result.dropItemCallback = _onDropItem;
-    return result;
-  }
-
-  void _onDropItem(_StoredMessage message) {
-    message.callback(null);
-  }
+  /// A mapping between a channel name and its associated [_Channel].
+  final Map<String, _Channel> _channels = <String, _Channel>{};
 
+  /// Adds a message (`data`) to the named channel buffer (`channel`).
+  ///
+  /// The `callback` argument is a closure that, when called, will send messages
+  /// back to the plugin.
+  ///
   /// Returns true on overflow.
-  bool push(String channel, ByteData? data, PlatformMessageResponseCallback callback) {
-    _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    if (queue == null) {
-      queue = _makeRingBuffer(kDefaultBufferSize);
-      _messages[channel] = queue;
-    }
-    final bool didOverflow = queue.push(_StoredMessage(data, callback));
-    if (didOverflow) {
-      // TODO(gaaclarke): Update this message to include instructions on how to resize
-      // the buffer once that is available to users and print in all engine builds
-      // after we verify that dropping messages isn't part of normal execution.
-      _printDebug('Overflow on channel: $channel. '
-                  'Messages on this channel are being discarded in FIFO fashion. '
-                  'The engine may not be running or you need to adjust '
-                  'the buffer size of the channel.');
-    }
-    return didOverflow;
-  }
-
-  /// Returns null on underflow.
-  _StoredMessage? _pop(String channel) {
-    final _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    final _StoredMessage? result = queue?.pop();
-    return result;
+  bool push(String name, ByteData? data, PlatformMessageResponseCallback callback) {
+    final _Channel channel = _channels.putIfAbsent(name, () => _Channel(kDefaultBufferSize));
+    return channel.push(_StoredMessage(data, callback));
   }
 
-  bool _isEmpty(String channel) {
-    final _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    return queue == null || queue.isEmpty;
+  /// Sets the listener for the specified channel.
+  ///
+  /// When there is a listener, messages are sent immediately.
+  ///
+  /// Each channel may have up to one listener set at a time. Setting
+  /// a new listener on a channel with an existing listener clears the
+  /// previous one.
+  ///
+  /// Callbacks are invoked in their own stack frame and
+  /// use the zone that was current when the callback was
+  /// registered.
+  ///
+  /// ## Draining
+  ///
+  /// If any messages were queued before the listener is added,
+  /// they are drained asynchronously after this method returns.
```

Comment:
Consider linking to the `drain` method at the end of the paragraph. A reader may not know what "drain" means.

Reply:
done